src/cr/cube/dimension.py

```diff
@@ -610,6 +612,50 @@ def _dimension_transforms_dict(self) -> Dict:
         """Copy of dimension transforms dictionary with shimmed `element_id`s"""
         return self._element_id_shim.shimmed_dimension_transforms_dict
 
+    @lazyproperty
+    def _element_data_format(self) -> Optional[Union[str, int]]:
+        """optional str format for datetimes or int number of decimals for numeric"""
+        return self._dimension_dict["references"].get("format", {}).get("data")
+
+    def _format_datetime_element_label(self, x: Any) -> str:
+        """returns str of x formatted for use in labels with desired date format"""
+        orig_format = self._incoming_element_datetime_format
+        out_format = self._element_data_format
+        if orig_format is None or out_format is None:
+            return str(x)
+        try:
+            return datetime.strptime(x, orig_format).strftime(out_format)
+        except ValueError:
+            return str(x)
+
+    def _format_element_label(self, x: Any) -> str:
+        """returns str of x formatted for use in labels"""
+        if self.dimension_type == DT.DATETIME:
+            return self._format_datetime_element_label(x)
+        else:  # --- TODO: Should round numeric values like the frontend does
+            return str(x)
+
+    @lazyproperty
+    def _incoming_element_datetime_format(self) -> Optional[str]:
+        """optional str of date formatting requested if datetime and is found"""
+        if self.dimension_type != DT.DATETIME:
+            return None
+
+        resolution = self._dimension_dict["type"]["subtype"].get("resolution")
+        return {
+            "Y": "%Y",
+            "Q": "%Y-%m",
+            "3M": "%Y-%m",
+            "M": "%Y-%m",
+            "W": "%Y-%m-%d",
+            "D": "%Y-%m-%d",
+            "h": "%Y-%m-%dT%H",
+            "m": "%Y-%m-%dT%H:%M",
+            "s": "%Y-%m-%dT%H:%M:%S",
+            "ms": "%Y-%m-%dT%H:%M:%S.%f",
+            "us": "%Y-%m-%dT%H:%M:%S.%f",
+        }.get(resolution)
```

Comment:
This supposes that resolution always exist in case of datetime. Is that true or do we need a fallback option in case of None value?

Reply:
`{...}.get(None)` is None, right? Do you want some way to make it more clear?

Reply:
my mistake ... that's fine